aea/runtime.py

```diff
@@ -261,165 +247,79 @@ def set_loop(self, loop: AbstractEventLoop) -> None:
+    async def run(self) -> None:
+        """Start multiplexeor task."""
+        terminal_state = RuntimeStates.error
+        try:
+            await self.run_runtime()
+        except StopRuntime as e:
+            self._state.set(RuntimeStates.stopping)
+            terminal_state = RuntimeStates.stopped
+            if e.reraise:
+                raise e.reraise
+        except (asyncio.CancelledError, CancelledError, KeyboardInterrupt):
+            self._state.set(RuntimeStates.stopping)
+            terminal_state = RuntimeStates.stopped
+        finally:
+            await self.stop_runtime()
+            self._state.set(terminal_state)
 
-        Start runtime asynchonously in own event loop.
+    async def stop_runtime(self) -> None:
         """
-        self.set_loop(self.loop)
+        Stop runtime coroutine.
 
-        logger.debug(f"Start runtime event loop {self.loop}: {id(self.loop)}")
-        self._task = self.loop.create_task(self.run_runtime())
+        Stop main loop.
+        Agent teardown.
+        Disconnect multiplexer.
+        """
+        self.main_loop.stop()
+        with suppress(StopRuntime):
+            await self.main_loop.wait_completed()
+        self._teardown()
 
-        try:
-            self.loop.run_until_complete(self._task)
-            logger.debug("Runtime loop stopped!")
-        except Exception:
-            logger.exception("Exception raised during runtime processing")
-            self._state.set(RuntimeStates.error)
-            raise
-        finally:
-            self._stopping_task = None
+        self.multiplexer.stop()
+        await self.multiplexer.wait_completed()
+        logger.debug("Runtime loop stopped!")
 
     async def run_runtime(self) -> None:
         """Run agent and starts multiplexer."""
         self._state.set(RuntimeStates.starting)
-        try:
-            await self._start_multiplexer()
-            await self._start_agent_loop()
-        except Exception:
-            logger.exception("AsyncRuntime exception during run:")
-            raise
-        finally:
-            if self._stopping_task and not self._stopping_task.done():
-                await self._stopping_task
-
-    async def _multiplexer_disconnect(self) -> None:
-        """Call multiplexer disconnect asynchronous way."""
-        await AsyncMultiplexer.disconnect(self.multiplexer)
+        await asyncio.gather(self._start_multiplexer(), self._start_agent_loop())
 
     async def _start_multiplexer(self) -> None:
         """Call multiplexer connect asynchronous way."""
         self.setup_multiplexer()
-        await AsyncMultiplexer.connect(self.multiplexer)
+        self.multiplexer.set_loop(self._loop)
+        self.multiplexer.start()
+        await self.multiplexer.wait_completed()
 
     async def _start_agent_loop(self) -> None:
         """Start agent main loop asynchronous way."""
         logger.debug("[{}]: Runtime started".format(self._agent.name))
```

Comment:
this should be moved above? In `run_runtime`?

Reply:
now, two methods to be awaited: loop and multiplexer